Generated grpc code cannot be compiled if a underscore is in the name of a method

### Describe the bug

I got a proto file which contains a rpc method with a underscore in the name

```
syntax = "proto3";

package mypackage;

option java_multiple_files = true;
option java_package = "mypackage";

service MyService {
  // Get info about the service
  rpc service_info(ServiceInfoRequest) returns (ServiceInfo) {}
}
message ServiceInfo {
  string software_version = 1;
}
message ServiceInfoRequest {
}
```

But the generated code is very inconsistend and cannot be compiled

e.g. in `MyServiceImplBase implements io.grpc.BindableService` its `getServiceinfoMethod` instead of `getServiceInfoMethod`

```
        @java.lang.Override
        public io.grpc.ServerServiceDefinition bindService() {
            return io.grpc.ServerServiceDefinition.builder(getServiceDescriptor()).addMethod(mypackage.MyServiceGrpc.getServiceinfoMethod(), asyncUnaryCall(new MethodHandlers<mypackage.ServiceInfoRequest, mypackage.ServiceInfo>(this, METHODID_SERVICE_INFO, compression))).build();
        }
```

causes

```
[ERROR]   generated-sources/grpc/mypackage/MutinyMyServiceGrpc.java:[73,117] cannot find symbol
[ERROR]   symbol:   method getServiceinfoMethod()
[ERROR]   location: class mypackage.MyServiceGrpc
```


and in ``MutinyMyServiceStub extends io.grpc.stub.AbstractStub<MutinyMyServiceStub> implements io.quarkus.grpc.MutinyStub`
it is `delegateStub::service_info` instead of `delegateStub::serviceInfo`

```
        public io.smallrye.mutiny.Uni<mypackage.ServiceInfo> service_info(mypackage.ServiceInfoRequest request) {
            return io.quarkus.grpc.stubs.ClientCalls.oneToOne(request, delegateStub::service_info);
        }
```

causes
```
[ERROR]   generated-sources/grpc/mypackage/MutinyMyServiceGrpc.java:[44,72] invalid method reference
[ERROR]   cannot find symbol
[ERROR]     symbol:   method service_info()
[ERROR]     location: class mypackage.MyServiceGrpc.MyServiceStub
```


### Expected behavior

Generated code is consistent and can be compiled

### Actual behavior

_No response_

### How to Reproduce?

Download Quarkus sample for grpc.

Add proto example file with name my-service.proto into /src/main/proto, see an example above.

Run mvn package

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

_No response_

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

--- a/extensions/grpc/deployment/src/test/proto/goodbyeworld.proto
+++ b/extensions/grpc/deployment/src/test/proto/goodbyeworld.proto
@@ -10,7 +10,7 @@ package goodbyeworld;
 // The farewell service definition.
 service Farewell {
     // Sends a farewell
-    rpc SayGoodbye (GoodbyeRequest) returns (GoodbyeReply) {}
+    rpc Say_goodbye (GoodbyeRequest) returns (GoodbyeReply) {}
 }
 
 // The request message containing the user's name.

--- a/extensions/grpc/protoc/pom.xml
+++ b/extensions/grpc/protoc/pom.xml
@@ -27,6 +27,11 @@
                 </exclusion>
             </exclusions>
         </dependency>
+        <dependency>
+           <groupId>org.junit.jupiter</groupId>
+           <artifactId>junit-jupiter</artifactId>
+           <scope>test</scope>
+        </dependency>
     </dependencies>
 
     <build>

--- a/extensions/grpc/protoc/src/main/java/io/quarkus/grpc/protoc/plugin/MutinyGrpcGenerator.java
+++ b/extensions/grpc/protoc/src/main/java/io/quarkus/grpc/protoc/plugin/MutinyGrpcGenerator.java
@@ -137,7 +137,7 @@ public class MutinyGrpcGenerator extends Generator {
     private MethodContext buildMethodContext(DescriptorProtos.MethodDescriptorProto methodProto, ProtoTypeMap typeMap,
             List<DescriptorProtos.SourceCodeInfo.Location> locations, int methodNumber) {
         MethodContext methodContext = new MethodContext();
-        methodContext.methodName = lowerCaseFirst(methodProto.getName());
+        methodContext.methodName = adaptMethodName(methodProto.getName());
         methodContext.inputType = typeMap.toJavaTypeName(methodProto.getInputType());
         methodContext.outputType = typeMap.toJavaTypeName(methodProto.getOutputType());
         methodContext.deprecated = methodProto.getOptions() != null && methodProto.getOptions().getDeprecated();
@@ -171,8 +171,96 @@ public class MutinyGrpcGenerator extends Generator {
         return methodContext;
     }
 
-    private String lowerCaseFirst(String s) {
-        return Character.toLowerCase(s.charAt(0)) + s.substring(1);
+    static String adaptMethodName(String name) {
+        if (name == null || name.isEmpty()) {
+            return name;
+        }
+        // We need to adjust the method name in the same way as the grpc-java compiler does:
+        // 1. Decapitalize the first letter (unlike JavaBeans "URL" becomes "uRL")
+        // 2. Replace an underscore and capitalize the following letter
+        // 3. Append the underscore for a java keyword
+        // (these rules are bit odd but we need to follow them unless we get rid of the grpc-java compiler)
+        // https://github.com/quarkusio/quarkus/issues/27170
+        StringBuilder ret = new StringBuilder();
+        ret.append(Character.toLowerCase(name.charAt(0)));
+        if (name.length() > 1) {
+            if (name.contains("_")) {
+                for (String str : name.substring(1).split("_")) {
+                    if (ret.length() == 1) {
+                        ret.append(str);
+                    } else {
+                        ret.append(Character.toUpperCase(str.charAt(0)));
+                        if (str.length() > 1) {
+                            ret.append(str.substring(1));
+                        }
+                    }
+                }
+            } else {
+                ret.append(name.substring(1));
+            }
+            if (isJavaKeyword(ret.toString())) {
+                ret.append("_");
+            }
+        }
+        return ret.toString();
+    }
+
+    private static boolean isJavaKeyword(String value) {
+        switch (value) {
+            case "public":
+            case "protected":
+            case "private":
+            case "abstract":
+            case "static":
+            case "final":
+            case "transient":
+            case "volatile":
+            case "synchronized":
+            case "native":
+            case "class":
+            case "interface":
+            case "extends":
+            case "package":
+            case "throws":
+            case "implements":
+            case "boolean":
+            case "byte":
+            case "char":
+            case "short":
+            case "int":
+            case "long":
+            case "float":
+            case "double":
+            case "void":
+            case "if":
+            case "else":
+            case "try":
+            case "catch":
+            case "finally":
+            case "do":
+            case "while":
+            case "for":
+            case "continue":
+            case "switch":
+            case "case":
+            case "default":
+            case "break":
+            case "throw":
+            case "return":
+            case "this":
+            case "new":
+            case "super":
+            case "import":
+            case "instanceof":
+            case "goto":
+            case "const":
+            case "null":
+            case "true":
+            case "false":
+                return true;
+            default:
+                return false;
+        }
     }
 
     private List<PluginProtos.CodeGeneratorResponse.File> generateFiles(List<ServiceContext> services) {

--- a/extensions/grpc/protoc/src/test/java/io/quarkus/grpc/protoc/plugin/MutinyGrpcGeneratorTest.java
+++ b/extensions/grpc/protoc/src/test/java/io/quarkus/grpc/protoc/plugin/MutinyGrpcGeneratorTest.java
@@ -1,0 +1,19 @@
+package io.quarkus.grpc.protoc.plugin;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import org.junit.jupiter.api.Test;
+
+public class MutinyGrpcGeneratorTest {
+
+    @Test
+    public void testAdaptMethodName() {
+        assertEquals("sayHello", MutinyGrpcGenerator.adaptMethodName("SayHello"));
+        assertEquals("sayHello", MutinyGrpcGenerator.adaptMethodName("Say_Hello"));
+        assertEquals("sayHello", MutinyGrpcGenerator.adaptMethodName("Say_hello"));
+        assertEquals("_SayHello", MutinyGrpcGenerator.adaptMethodName("_Say_hello"));
+        assertEquals("sayHelloAndBye", MutinyGrpcGenerator.adaptMethodName("Say_Hello_and_Bye"));
+        assertEquals("return_", MutinyGrpcGenerator.adaptMethodName("return"));
+    }
+
+}

